File: Software/Chapter 14/vb/Listing14-4/Listing14-4/Listing14-4.vb
Imports System
Imports System.Data
Imports System.Data.SqlClient
Imports System.Data.SqlTypes
Imports Microsoft.SqlServer.Server

Namespace APress.Samples
    <Serializable()> _
    <SqlUserDefinedAggregate(Format.Native)> _
    Public Structure Range

        Private min As SqlDouble
        Private max As SqlDouble

        Public Sub Init()
            Me.min = SqlDouble.Null
            Me.max = SqlDouble.Null
        End Sub

        Public Sub Accumulate(ByVal value As SqlDouble)
            If Not value.IsNull Then
                If Me.min.IsNull OrElse value < Me.min Then
                    Me.min = value
                End If
                If Me.max.IsNull OrElse value > Me.max Then
                    Me.max = value
                End If
            End If
        End Sub

        Public Sub Merge(ByVal tempRange As Range)
            If (Me.min.IsNull OrElse tempRange.min < Me.min) Then
                Me.min = tempRange.min
            End If
            If (Me.min.IsNull OrElse tempRange.max > Me.max) Then
                Me.max = tempRange.max
            End If
        End Sub

        Public Function Terminate() As SqlDouble
            Dim result As SqlDouble = SqlDouble.Null
            If Not (Me.min.IsNull OrElse Me.max.IsNull) Then
                result = Me.max - Me.min
            End If
            Return result
        End Function
    End Structure
End Namespace

File: Software/Chapter 14/cs/Listing14-4/Listing14-4/Listing14-4.cs
using System;
using System.Data;
using System.Data.SqlClient;
using System.Data.SqlTypes;
using Microsoft.SqlServer.Server;

namespace APress.Samples
{
    [Serializable]
    [Microsoft.SqlServer.Server.SqlUserDefinedAggregate(Format.Native)]
    public struct Range
    {
        private SqlDouble min;
        private SqlDouble max;

        public void Init()
        {
            this.min = SqlDouble.Null;
            this.max = SqlDouble.Null;
        }

        public void Accumulate(SqlDouble Value)
        {
            if (!Value.IsNull)
            {
                if (this.min.IsNull || Value < this.min)
                    this.min = Value;
                if (this.max.IsNull || Value > this.max)
                    this.max = Value;
            }
        }

        public void Merge(Range tempRange)
        {
            if (this.min.IsNull || tempRange.min < this.min)
                this.min = tempRange.min;
            if (this.min.IsNull || tempRange.max > this.max)
                this.max = tempRange.max;
        }

        public SqlDouble Terminate()
        {
            SqlDouble result = new SqlDouble();
            result = SqlDouble.Null;
            if (!this.min.IsNull && !this.max.IsNull)
                result = this.max - this.min;
            return result;
        }
    };
}